04-cs-fundamentals/internship/binary-search-trees.md

```diff
@@ -1,23 +1,253 @@
+
+![bst.png](images/bst.png)
+
+Notice that nodes 61 and 77, both form subtrees of the larger tree data structure.  This enables you to use **recursion** on most tree methods because each child node forms a subtree.
+
+**Exercise**
+
+Given the following values draw a binary search tree:
+
+"Ringo", "John", "Paul", "George"
+
+## Binary Search Tree Node
+
+Just like a Linked List node, a tree node has a data field, and links to other nodes.  However instead of linking to previous and next nodes, the links are to nodes less than and greater than the value in the data field.
+
+![Binary Search Tree Example](images/bst1.png)
+
+The Node class for a Binary Search Tree (BST) could look like this:
+
+```ruby
+class TreeNode
+  attr_reader :data
+  attr_accessor :left, :right
+
+  def initialize(data)
+    @data = data
+    @left = @right = nil
+  end
+end
+```
+
+These nodes are arranged with one node known as the **root** which is where the tree begins.  New nodes as they are added are placed to the left, if they are less than the current node, and to the right if they are greater than the current node.  This is a natually recursive process.
+
+```
+Method add:
+  if the root is nil set the root to be a new node with the given value and return the node.
+
+  Otherwise 
+    if the value is less than or equal to the current root's value, make root's left be the result of calling add on root's left.
+    otherwise make root's right be the result of calling add on root's right.
+```
+
+You can observe this in the [Binary Tree Visualizer](http://btv.melezinek.cz/binary-search-tree.html)
+
+
+## The Tree Class
+
+You can build a `Tree` class in a manner similar to how we created a `LinkedList` class.
+
+```ruby
+class Tree
+  attr_reader :root
+
+  def initialize
+    @root = nil
+  end
+
+  # Tree methods go here...
+end
+```
+
+## Finding a Value
+
+You can try to search to find a value in a Binary Search Tree Like this:
+
+```
+If root is nil return nil
+If the root equals the value being searched for return the root's data
+
+If the value is less than the root return search on root's left side
+If the value is greater than the root return search on root's right side
+```
+
+**Exercise** 
+
+Try this out on the [Binary Tree Visualizer](http://btv.melezinek.cz/binary-search-tree.html).
+
+**Question**:  What's the worst-case for finding a value in the tree?
+
+**Question**:  What's the best-case for finding a value in the tree?
+
+**Question**:  What would you guess is the average case for finding a node in a tree?
+
+We will answer these questions in a bit.
+
+## Balanced Trees & Unbalanced Trees
+
+A tree is considered **balanced** if the height of it's left and right subtree differ by at most 1.  In this way the nodes in the tree must be spread fairly evenly.
```

Comment:
```suggestion
A tree is considered **balanced** if the height of its left and right subtree differ by at most 1.  In this way the nodes in the tree must be spread fairly evenly.
```

It's is short for "it is", its is the possessive.

Reply:
I'm not sure this is a good definition of "balanced" - I think that depends on your balancing algorithm.

http://user.it.uu.se/~justin/Teaching/PK2/Slides/AVLtrees.pdf

Reply:
Hmmm, now that I think about that more I'm not sure. I think it's a complex discussion either way.

Reply:
I definitely think balanced-ness needs to be defined recursively.

This is not a balanced tree:
```
  /\
 /  \
/    \
       \
```